ArmPkg/Include/AsmMacroIoLibV8.h

```diff
@@ -12,42 +12,138 @@
+#define NUM(x)         CATSTR(HASH,x)
+#define HASH  #
+// MU_CHANGE [END] - ARM64 VS change
+
 // CurrentEL : 0xC = EL3; 8 = EL2; 4 = EL1
 // This only selects between EL1 and EL2, else we die.
 // Provide the Macro with a safe temp xreg to use.
+#if !defined (_MSC_VER)   // MU_CHANGE - ARM64 VS change
 #define EL1_OR_EL2(SAFE_XREG)        \
         mrs    SAFE_XREG, CurrentEL ;\
         cmp    SAFE_XREG, #0x8      ;\
         b.gt   .                    ;\
         b.eq   2f                   ;\
         cbnz   SAFE_XREG, 1f        ;\
         b      .                    ;// We should never get here
-
+// MU_CHANGE [BEGIN] - ARM64 VS change
+#else
+#define EL1_OR_EL2(SAFE_XREG)        \
+        mrs    SAFE_XREG, CurrentEL __CR__\
+        cmp    SAFE_XREG, NUM(0x8)  __CR__\
+6
+bgt    %b6                  __CR__ \
+  beq    %f2                  __CR__ \
+  cbnz   SAFE_XREG, NUM (0x4)  __CR__ \
+  5                                   __CR__ \
+  bne    %b5                       // We should never get here
+#endif
+// EL1 code starts here
+// MU_CHANGE [END] - ARM64 VS change
 // CurrentEL : 0xC = EL3; 8 = EL2; 4 = EL1
 // This only selects between EL1 and EL2 and EL3, else we die.
 // Provide the Macro with a safe temp xreg to use.
+#if !defined (_MSC_VER)
 #define EL1_OR_EL2_OR_EL3(SAFE_XREG) \
         mrs    SAFE_XREG, CurrentEL ;\
         cmp    SAFE_XREG, #0x8      ;\
         b.gt   3f                   ;\
         b.eq   2f                   ;\
         cbnz   SAFE_XREG, 1f        ;\
         b      .                    ;// We should never get here
+// MU_CHANGE [BEGIN] - ARM64 VS change
+#else
+#define EL1_OR_EL2_OR_EL3(SAFE_XREG) \
+        mrs    SAFE_XREG, CurrentEL  __CR__\
+        cmp    SAFE_XREG, NUM(0x8)   __CR__\
+                bgt    %f3                   __CR__\
+        beq    %f2                   __CR__\
+        cbnz   SAFE_XREG, %f1        __CR__\
+5                                    __CR__\
+        bne    %b5                // We should never get here
+#endif
+// EL1 code starts here
+
+#if defined (_MSC_VER)
+
+// MU_CHANGE - add
+#define LoadConstantToReg(Data, Reg) \
+  ldr Reg, =Data
+
+#elif defined (__clang__)
+
+// load x0 with _Data
+#define LoadConstant(_Data)              \
+  ldr  x0, 1f                          ; \
+  b    2f                              ; \
+.align(8)                              ; \
+1:                                       \
+  .8byte (_Data)                       ; \
+2:
+
+// load _Reg with _Data
+#define LoadConstantToReg(_Data, _Reg)    \
+  ldr  _Reg, 1f                         ; \
+  b    2f                               ; \
+.align(8)                               ; \
+1:                                        \
+  .8byte (_Data)                        ; \
+2:
+
+#elif defined (__GNUC__)
+
+#define LoadConstant(Data) \
+  ldr  x0, =Data
+
+#define LoadConstantToReg(Data, Reg) \
+  ldr  Reg, =Data
+
+#endif // __GNUC__
 
+#if !defined (_MSC_VER)

```

Comment:
MU_CHANGE tag

Reply:
This one is within a multi-line MU_CHANGE block.